lib/auth/accountrecovery.go

```diff
@@ -373,6 +376,81 @@ func (s *Server) recordFailedRecoveryAttempt(ctx context.Context, username strin
 	return lockUntil, maxedAttempts, nil
 }
 
+// CompleteAccountRecovery implements AuthService.CompleteAccountRecovery.
+func (s *Server) CompleteAccountRecovery(ctx context.Context, req *proto.CompleteAccountRecoveryRequest) error {
+	if err := s.isAccountRecoveryAllowed(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	approvedToken, err := s.GetUserToken(ctx, req.GetRecoveryApprovedTokenID())
+	if err != nil {
+		log.Error(trace.DebugReport(err))
+		return trace.AccessDenied(completeRecoveryGenericErrMsg)
+	}
+
+	if err := s.verifyUserToken(approvedToken, UserTokenTypeRecoveryApproved); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check that the correct auth credential is being recovered before setting a new one.
+	switch req.GetNewAuthnCred().(type) {
+	case *proto.CompleteAccountRecoveryRequest_NewPassword:
+		if approvedToken.GetUsage() == types.UserTokenUsage_USER_TOKEN_RECOVER_MFA {
+			log.Debugf("Failed to recover account, expected new mfa register response, but received a new password.")
+			return trace.AccessDenied(completeRecoveryGenericErrMsg)
+		}
+
+		if err := services.VerifyPassword(req.GetNewPassword()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := s.UpsertPassword(approvedToken.GetUser(), req.GetNewPassword()); err != nil {
+			log.Error(trace.DebugReport(err))
+			return trace.AccessDenied(completeRecoveryGenericErrMsg)
+		}
+
+	case *proto.CompleteAccountRecoveryRequest_NewMFAResponse:
+		if approvedToken.GetUsage() == types.UserTokenUsage_USER_TOKEN_RECOVER_PASSWORD {
+			log.Debugf("Failed to recover account, expected new password, but received a new mfa register response.")
+			return trace.AccessDenied(completeRecoveryGenericErrMsg)
+		}
+
+		err := s.verifyMFARespAndAddDeviceWithToken(ctx, req.GetNewMFAResponse(), approvedToken, req.GetNewDeviceName())
+		switch {
+		case trace.IsAlreadyExists(err):
+			// Return a shorter friendlier message.
+			return trace.AlreadyExists("MFA device with name %q already exists", req.GetNewDeviceName())
+		case err != nil:
+			return trace.Wrap(err)
+		}
+
+	default:
+		return trace.AccessDenied("unsupported authentication method")
+	}
+
+	// Check and remove user locks so user can immediately sign in after finishing recovering.
+	user, err := s.GetUser(approvedToken.GetUser(), false)
```

Comment:
```suggestion
	user, err := s.GetUser(approvedToken.GetUser(), false /* withSecrets */)
```

(assuming I got that right, of course)